#!/usr/bin/env bash
#PBS -q omp
#PBS -l ncpus=21
#PBS -l mem=40gb
#PBS -l walltime=24:00:00
#PBS -J 1-58
#PBS -N 03b_star_mapping


#declare variables used in the script
DATADIRECTORY=
OUT=
READFILES=$OUT/evotox/03_trimmed
OUTDIR=$OUT/evotox/04_mapped
GENOMEFOLDER=$OUT/evotox/01_info_files
GENOMEFASTA=$GENOMEFOLDER/GCF_020424385.1_ASM2042438v2_genomic.fna
GFF3=$GENOMEFOLDER/GCF_020424385.1_ASM2042438v2_genomic.gff
LOG=$DATADIRECTORY/98_log/03_star_mapping

TMPDIR=$OUT
TMP=$OUT

SAMBAMBA=Path/to/conda/bin
STAR=Path/to/conda/bin


mkdir -p $OUTDIR
mkdir -p $LOG

# Assign array nb to file
NAME=$(cat $DATADIRECTORY/00_scripts/01_base_fastp.txt | awk "NR==${PBS_ARRAY_INDEX}")

#save array table correspondance
echo -e ${PBS_ARRAY_INDEX}"\t"${NAME} >> $DATADIRECTORY/00_scripts/03b_star_array.txt

$STAR
STAR --runThreadN 20 --genomeDir $GENOMEFOLDER --readFilesIn $READFILES/${NAME}_R1_paired.fastq.gz $READFILES/${NAME}_R2_paired.fastq.gz --readFilesCommand gunzip -c --sjdbGTFfile $GFF3 --sjdbOverhang 149 --outFileNamePrefix $OUTDIR/${NAME}_ --outSAMattrRGline ID:${NAME} SM:${NAME} &> $LOG/03b_star_GFF_${NAME}.log

conda deactivate


$SAMBAMBA
cd $OUTDIR
sambamba view -t 20 -S -f bam ${NAME}_Aligned.out.sam &> ${NAME}_Aligned.out.bam
